### PR TITLE
Feat/word export

### LINKED ISSUE #27: Word export format
The final aim of the project is to generate a competency document with as little human intervention as possible. To achieve this goal, it is a good idea to implement an exporter for Word files that can export achieved KPIs in the form of a table.

### PR DESCRIPTION
**Pull request by NealGeilen**
Adding the functionality to the project for exporting to "Word" files aka ".docx"

This is mostly done as a test to see what it takes to create word documents within .NET. 
I wanted to test this so that there is a base when we want to start the development of a "competence document generator"

### CODE REVIEW

**Comment by Typiqally** [comment]
Nice work, I'll take a look when you are ready for review :D. Should really refactor the code again some day, It's still a bit messy for my liking.

EDIT: To clarify, I'm talking about the general codebase, not your commits.

**Comment by NealGeilen** [comment]
I was thinking, is it still needed to have a pdf export format? #27

I just don't think that there is a need for it. Especially looking at the fact that you can't edit a pdf document. With this a user can only copy the contents of the document. And you can already do that with the other export formats.

**Comment by Typiqally** [comment]
> I was thinking, is it still needed to have a pdf export format? #27
> 
> I just don't think that there is a need for it. Especially looking at the fact that you can't edit a pdf document. With this a user can only copy the contents of the document. And you can already do that with the other export formats. 

That's fair, you can remove it from the issue if you want. Do you think it is also possible to include other "Word" formats such as otd, for Linux users?

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                {
+                    var assignmentIds = module.Collection.OutcomeResults
+                        .Where(o => o.Link.Outcome == outcomeId)
+                        .Select(static o => o.Link.Assignment)
+                        .ToArray();
+
+                    if (assignmentIds.Any())
+                    {
+                        var row = table.InsertRow();
+                        row.Cells[ 0 ].Paragraphs[ 0 ].Append( outcome.Title + " " +  ShortDescription(ConvertHtmlToRaw(outcome.Description)) );
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: Can you reformat the spaces around the array index and parenthesis? They do not conform to the coding style Also for the rest of this class.

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static string ConvertHtmlToRaw(string html)
+    {
+        var raw = Regex.Replace(html, "<.*?>", " ");
+        var trimmed = Regex.Replace(raw, @"\s\s+", " ");
+
+        return trimmed;
+    }
+
+    private string OutcomeToText(double? result)
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: Maybe we can change this into a public helper function, maybe an extension method for OutcomeResult (e.g., outcomeResult.toUsgo() or something similar)

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                } 
+                var par = document.InsertParagraph( module.Name );
+                par.FontSize(24);
+                par.InsertTableAfterSelf(table).InsertPageBreakAfterSelf();
+            }
+            document.Save();
+        }
+    }
+    
+    private static string ShortDescription(string description)
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: We should also start moving these methods to a helper function of sorts.

**Review by Typiqally — CHANGES_REQUESTED** [review]
Nice work, left some small review comments. I was also wondering if you spent time on Linux compatible formats such as OTD?

**Comment by NealGeilen** [comment]
> I was also wondering if you spent time on Linux compatible formats such as OTD?

At this moment I cant find a .NET lib that supports this format.

**Comment by Typiqally** [comment]
> > I was also wondering if you spent time on Linux compatible formats such as OTD?
> 
> At this moment I cant find a .NET lib that supports this format.

That's fine, I believe many Word editing tools support the MS Word format on Linux anyway.

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 4]
```diff
@@ -53,7 +52,7 @@ public void Export(IEnumerable<Module> modules, string format)
 
                 if (assignmentIds.Any())
                 {
-                    worksheet.Cells[index, 0] = new Cell(outcome.Title + " " +  ShortDescription(ConvertHtmlToRaw(outcome.Description)));
+                    worksheet.Cells[index, 0] = new Cell(outcome.Title + " " +  outcome.ShortDescription());
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: Can you use string interpolation here?

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    public WordModuleExporter(IOptions<ExportOptions> options)
+    {
+        _options = options.Value;
+    }
+
+    public IEnumerable<string> Formats { get; } = new[] { "word" };
+
+    public void Export(IEnumerable<Module> modules, string format)
+    {
+        using (DocX document = DocX.Create(_options.FormattedOutputName+".docx"))
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: String interpolation here as well

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            foreach (var module in modules.Where(static m => m.Collection.OutcomeResults.Any())) 
+            {
+
+                var links = module.Collection.Links;
+                var alignments = links.AlignmentsDictionary;
+                var outcomes = links.OutcomesDictionary;
+                
+                var table = document.AddTable( 1, 3);
+
+                table.Rows[ 0 ].Cells[ 0 ].Paragraphs[ 0 ].Append("KPI");
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: Maybe for convenience, just format this document using the following shortcut

- Rider: CTRL + ALT + L
- Visual Studio: CTRL + K, CTRL + F
- VS Code: ALT + Shift + F

**Review by Typiqally — CHANGES_REQUESTED** [review]
I'm really liking the helper functions, nicely done, still have some minor changes requests

**Pull request merged**